File: 02-tech/01-programming-fundamentals/13-exam/02-softuni-course-planning/SoftuniCoursePlanning.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace _02_softuni_course_planning
{
    class SoftuniCoursePlanning
    {
        static void Main()
        {
            #if DEBUG
            Console.SetIn(new StreamReader("tests/test2.txt"));
            #endif

            var schedule = Console.ReadLine().Split(new string[] { ", " }, StringSplitOptions.RemoveEmptyEntries).ToList();
            var lessons = new List<string>();
            var lessonHasExercise = new Dictionary<string, bool>();

            foreach (var lesson in schedule)
            {
                if (lessonHasExercise.ContainsKey(lesson) == false)
                {
                    lessonHasExercise[lesson] = false;
                    lessons.Add(lesson);
                }
            }

            var input = Console.ReadLine();
            while (input != "course start")
            {
                var data = input.Split(':');
                var command = data[0];
                var lesson = data[1];

                if (command == "Add")
                {
                    if (lessonHasExercise.ContainsKey(lesson) == false)
                    {
                        lessons.Add(lesson);
                        lessonHasExercise[lesson] = false;
                    }
                }
                else if (command == "Insert")
                {
                    var index = int.Parse(data[2]);
                    // bool indexIsValid = 0 <= index && index < schedule.Count;
                    if (lessonHasExercise.ContainsKey(lesson) == false)
                    {
                        lessons.Insert(index, lesson);
                        lessonHasExercise[lesson] = false;
                    }
                }
                else if (command == "Remove")
                {
                    lessons.Remove(lesson);
                    lessonHasExercise.Remove(lesson);
                }
                else if (command == "Swap")
                {
                    var secondLesson = data[2];
                    var indexOfFirstLesson = lessons.IndexOf(lesson);
                    var indexOfSecondLesson = lessons.IndexOf(secondLesson);

                    if (indexOfFirstLesson >= 0 && indexOfSecondLesson >= 0)
                    {
                        var temp = lessons[indexOfFirstLesson];
                        lessons[indexOfFirstLesson] = lessons[indexOfSecondLesson];
                        lessons[indexOfSecondLesson] = temp;
                    }
                }
                else if (command == "Exercise")
                {
                    if (lessonHasExercise.ContainsKey(lesson) == false)
                    {
                        lessonHasExercise[lesson] = false;
                        lessons.Add(lesson);
                    }

                    lessonHasExercise[lesson] = true;
                }

                input = Console.ReadLine();
            }

            for (int current = 0, count = 1; current < lessons.Count; current++)
            {
                Console.WriteLine($"{count++}.{lessons[current]}");
                if (lessonHasExercise.ContainsKey(lessons[current]) && lessonHasExercise[lessons[current]])
                {
                    Console.WriteLine($"{count++}.{lessons[current]}-Exercise");
                }
            }
        }
    }
}

File: 02-tech/01-programming-fundamentals/13-exam/02-softuni-course-planning/softuni-course-planning.js
// jshint esversion:6
function solve(commands)
{
    // 'use strict';
    const lessons = commands.shift().split(', ');
    const lessonsWithExercise = new Map();
    for (let lesson of lessons)
    {
        lessonsWithExercise.set(lesson, false);
    }

    commands.pop();
    for (let command of commands)
    {
        const data = command.split(':');
        const commandType = data[0];
        const lesson = data[1];

        if (commandType == 'Add')
        {
            if (lessonsWithExercise.has(lesson) == false)
            {
                lessons.push(lesson);
                lessonsWithExercise.set(lesson, false);
            }
        }
        else if (commandType == 'Insert')
        {
            const index = +data[2];

            if (lessonsWithExercise.has(lesson) == false)
            {
                lessons.splice(index, 0, lesson);
                lessonsWithExercise.set(lesson, false);
            }
        }
        else if (commandType == 'Remove')
        {
            const indexOfLesson = lessons.indexOf(lesson);

            if (indexOfLesson >= 0)
            {
                lessons.splice(indexOfLesson, 1);
                lessonsWithExercise.delete(lesson);
            }
        }
        else if (commandType == 'Swap')
        {
            const firstLesson = lesson;
            const secondLesson = data[2];
            const indexOfFirstLesson = lessons.indexOf(firstLesson);
            const indexOfSecondLesson = lessons.indexOf(secondLesson);

            if (indexOfFirstLesson >= 0 && indexOfSecondLesson >= 0)
            {
                lessons[indexOfFirstLesson] = secondLesson;
                lessons[indexOfSecondLesson] = firstLesson;
            }
        }
        else if (commandType == 'Exercise')
        {
            if (lessonsWithExercise.has(lesson) == false)
            {
                lessons.push(lesson);
            }

            lessonsWithExercise.set(lesson, true);
        }
    }

    for (let current = 0, course = 1; current < lessons.length; ++current)
    {
        console.log(`${course++}.${lessons[current]}`);
        if (lessonsWithExercise.get(lessons[current]))
        {
            console.log(`${course++}.${lessons[current]}-Exercise`);
        }
    }
}

// solve([
//     'Data Types, Objects, Lists, Exercise: Databases, Exercise: Lists',
//     'Add:Databases',
//     'Insert:Arrays:0',
//     'Remove:Lists',
//     'Exercise:Databases',
//     'Exercise:Databases',
//     'Exercise:Databases',
//     'Exercise:Lists',
//     'course start'
// ]);

// solve([
//     'Data Types, Objects, Lists',
//     'Add:Databases',
//     'Insert:Arrays: 0',
//     'Remove:Lists',
//     'course start'
// ]);

// solve([
//     'Arrays, Lists, Methods',
//     'Swap:Arrays:Methods',
//     'Exercise:Databases',
//     'Swap:Lists:Databases',
//     'Insert:Arrays:0',
//     'course start'
// ]);